File: src/lib/intelligence/finalDecision.ts
import type { DayContext, UserProfile } from '../engineFacade';
import { calculateRuleScore } from './ruleScoring';
import { getOptionalModelScore } from './enhancer';
import type { IntelligenceDecision } from './types';

const RULE_WEIGHT = 0.7;
const MODEL_WEIGHT = 0.3;

export async function buildIntelligenceDecision(profile: UserProfile, day: DayContext, message: string): Promise<IntelligenceDecision> {
  const base = calculateRuleScore(profile, day, message);

  try {
    const modelScore = await Promise.race([
      getOptionalModelScore({ profile, day, message }),
      new Promise<null>((resolve) => setTimeout(() => resolve(null), 500)),
    ]);

    if (modelScore === null || Number.isNaN(modelScore)) {
      return base;
    }

    const finalScore = (base.ruleScore * RULE_WEIGHT + modelScore * MODEL_WEIGHT) / (RULE_WEIGHT + MODEL_WEIGHT);

    return {
      ...base,
      modelScore,
      finalScore,
      fallbackUsed: false,
      priority:
        finalScore >= 0.75 ? 'urgent' :
        finalScore >= 0.55 ? 'high' :
        finalScore >= 0.35 ? 'medium' : 'low',
    };
  } catch {
    return base;
  }
}

File: android/app/src/main/java/com/jainkundali/app/domain/intelligence/FinalDecision.kt
package com.jainkundali.app.domain.intelligence

import com.jainkundali.app.domain.models.DayContext
import com.jainkundali.app.domain.models.UserProfile
import kotlinx.coroutines.withTimeoutOrNull

/**
 * Combines the rule score with an optional model score (REFERENCE.md §3):
 *
 *     finalScore = (ruleScore × 0.7 + modelScore × 0.3) / (0.7 + 0.3)
 *
 * The model call is bounded by a hard 500 ms timeout; on timeout / null / failure we fall back to
 * the rule-only decision so the app never blocks or fails because of the optional layer.
 */
object FinalDecision {

    private const val RULE_WEIGHT = 0.7
    private const val MODEL_WEIGHT = 0.3
    private const val MODEL_TIMEOUT_MS = 500L

    suspend fun build(profile: UserProfile, day: DayContext, message: String): IntelligenceDecision {
        val base = RuleScoring.calculate(profile, day, message)

        val modelScore: Double? = try {
            withTimeoutOrNull(MODEL_TIMEOUT_MS) {
                Enhancer.optionalModelScore(profile, day, message)
            }
        } catch (e: Exception) {
            null
        }

        if (modelScore == null || modelScore.isNaN()) {
            return base
        }

        val finalScore =
            (base.ruleScore * RULE_WEIGHT + modelScore * MODEL_WEIGHT) / (RULE_WEIGHT + MODEL_WEIGHT)

        return base.copy(
            modelScore = modelScore,
            finalScore = finalScore,
            fallbackUsed = false,
            priority = RuleScoring.priorityFor(finalScore)
        )
    }
}
